### ISSUE TITLE
Check end of dataset using COCOReader pipe

### ISSUE THREAD

**Issue by dariodematties** [NVIDIA/DALI#2407]
Hi, I am using a cocoreader pipe to read images from MSCOCO.

It returns images, bboxes and labels.

I want to keep the output format (no pytorch but dali tensorlists) from this pipe since I have to use such output to feed another pipe.

How can I detect the end of the dataset (or the last batch) in order to take control of this pipe running iteratively in a loop.

Thanks!

**Comment by JanuszL** [comment]
Hi,
You need to manually count samples that you consumed from DALI and compare it with the dataset size you can query using [`reader_meta` method](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/pipeline.html#nvidia.dali.pipeline.Pipeline.reader_meta) from the pipeline. For multi-GPU training and shards sizes that are not a multiply the batch size, it can be tricky. The logic that iterators have to handle that properly can be found in [this file](https://github.com/NVIDIA/DALI/blob/master/dali/python/nvidia/dali/plugin/base_iterator.py).
I would recommend using `pad_last_batch` in the reader to make the shards equal and a multiply of the batch size - it should make the math simpler.

**Comment by dariodematties** [comment]
Thank you @JanuszL, I do pipe1.reader_meta(). With 2 processes I get:

```
pipe1 dataset information from rank 1 is {'epoch_size': 40504, 'epoch_size_padded': 40504, 'number_of_shards': 2, 'shard_id': 1, 'pad_last_batch': 1, 'stick_to_shard': 0}

pipe1 dataset information from rank 0 is {'epoch_size': 40504, 'epoch_size_padded': 40504, 'number_of_shards': 2, 'shard_id': 0, 'pad_last_batch': 1, 'stick_to_shard': 0}
```

As I can see, epoch_size is the total size of the dataset and is replicated in each process. On the other hand I do not understand the function of stick_to_shard. I enable it but the situation does not seem to change

From the documentation about CocoReader op:
stick_to_shard (bool, optional, default = False) – Whether reader should stick to given data shard instead of going through the whole dataset. When decoder caching is used, it reduces significantly the amount of data to be cached, but could affect accuracy in some cases

I do not quite understand the semantics behind this. As I understand it, I think each process will take a different batch from the dataset, in my case with 2 processes, process 0 will bring batches 0,2,4,... while process 1 will bring batches 1,3,5,...
I think this is the default behavior (stick_to_shard=False)
Yet, I do not understand what stick_to_shard actually means when you enable it

**Comment by JanuszL** [comment]
> As I can see, epoch_size is the total size of the dataset and is replicated in each process.

That is good, after all, all of them are working on the same data set. The only difference you may see if when you enable `pad_last_batch` and change the batch size/number of shards `epoch_size_padded` will adjust to that.

> On the other hand I do not understand the function of stick_to_shard. I enable it but the situation does not seem to change

The only thing that is affected is which shards are seen by a given GPU change epoch to epoch. By default each DALI process after each epoch moves to the next shard. So in your cases, GPU 0 will create batches from the shard 0 during the first epoch, shard 1 during the second, shard 0 during the third, and so on. If you enable this option all GPUs will be assigned to the starting shard for the whole training lifetime. I don't think you need to change the default behavior in your case.

**Comment by dariodematties** [comment]
Thank you @JanuszL, you are clarifying my thoughts about this topic since I now realize I didn't understand what a shard actually means


In [this link](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/multigpu.html#Sharding) I could find some guidance regarding to sharding in DALI. So, if I understood it correctly, a shard is a section of the dataset devoted to a process, isn't it?

And since I map one process per GPU I also have one GPU per shard

Therefore, in order to make each process receive all the dataset, there is a kind of rotation, epoch after epoch, in which different GPUs manage different shards.

Ok, so far so good.

In one node with a dual GPU I have the following scenario

```
GPU:0
local_rank = device_id = shard_id = 0
num_shards = world_size = 2

GPU:1
local_rank = device_id = shard_id = 1
num_shards = world_size = 2

```

But, what if I have 2 nodes with 2 GPUs each?

I will give it a try and correct me if I am wrong 

```
Node 0:

GPU:0
local_rank = device_id = shard_id = 0
num_shards = world_size = 4

GPU:1
local_rank = device_id = shard_id = 1
num_shards = world_size = 4


Node 1:

GPU:0
local_rank = device_id = shard_id = 2
num_shards = world_size = 4

GPU:1
local_rank = device_id = shard_id = 3
num_shards = world_size = 4

```

Excuse my basic questions please and thank you for your patience :)

**Comment by JanuszL** [comment]
Yes, @dariodematties you are correct. Shard is a continuous, nonoverlapping piece of data set dedicated to a given GPU. Thanks to shading DALI doesn't need any global coordination between processes/nodes regarding which GPU gets which sample, as each of works on a separate piece of the dataset.
And yes, with multi-node training each DALI pipeline instance should get a shard equal to global_rank id.

**Comment by dariodematties** [comment]
Thank you very much for all your help @JanuszL! : )
